api/src/main/java/org/apache/iceberg/types/ReassignIds.java

```diff
@@ -49,39 +54,53 @@ public Type struct(Types.StructType struct, Iterable<Type> fieldTypes) {
     Preconditions.checkArgument(sourceType.isStructType(), "Not a struct: %s", sourceType);
 
     Types.StructType sourceStruct = sourceType.asStructType();
+
     List<Types.NestedField> fields = struct.fields();
     int length = fields.size();
 
     List<Type> types = Lists.newArrayList(fieldTypes);
     List<Types.NestedField> newFields = Lists.newArrayListWithExpectedSize(length);
     for (int i = 0; i < length; i += 1) {
       Types.NestedField field = fields.get(i);
-      int sourceFieldId = sourceStruct.field(field.name()).fieldId();
+      Types.NestedField sourceField = sourceStruct.field(field.name());
+      int fieldId = fieldId(sourceField);
       if (field.isRequired()) {
-        newFields.add(Types.NestedField.required(sourceFieldId, field.name(), types.get(i), field.doc()));
+        newFields.add(Types.NestedField.required(fieldId, field.name(), types.get(i), field.doc()));
       } else {
-        newFields.add(Types.NestedField.optional(sourceFieldId, field.name(), types.get(i), field.doc()));
+        newFields.add(Types.NestedField.optional(fieldId, field.name(), types.get(i), field.doc()));
       }
     }
 
     return Types.StructType.of(newFields);
   }
 
+  private int fieldId(Types.NestedField sourceField) {
+    if (sourceField == null || newField) {
```

Comment:
What is the importance of the newField check? Is it not safe to assume that when `sourceField == null` it is a new field?